### PR TITLE
Alignat math environment

### PR DESCRIPTION
**Pull request by dgrigonis**
Removed unrelated edits and all that is left is Alignat environment with changes that were needed to implement it or allow for its nice behavior.

Implemented:
* Implemented ```alignat ``` environment for aligned equations
* Added ```_star_latex_name``` for LatexObject - (Many environments and commands can use this)

Edited / fixed:
* Edited ```Math``` class so that it could act as raw math container without enclosing math brackets
* Edited ```Math``` class has arguments for LatexMathParser and NoEscape() map for strings

### CODE REVIEW

**Comment by JelteF** [comment]
Hi thanks for the pull request. If possible could you try to split it into a couple of smaller pull requests, with seperated changes (for instance a seperate one for numbering bug for sections). That makes the changes much much more easy for me to review.

As for the parser I'd rather keep the goal of pylatex simple: generating latex code. Parsers tend to get complex in no time.

**Comment by dgrigonis** [comment]
Refactored unrelated changes

**Comment by JelteF** [comment]
Thanks that helps a lot. I'll try to go read this one in detail in the near future, but for now two points:
- Could you add changelog entries for your changes
- What's the reason that you didn't use the normal opt out for escaping? https://github.com/JelteF/PyLaTeX/blob/master/pylatex/base_classes/latex_object.py#L50-L66

**Comment by dgrigonis** [comment]
I would imagine I would have to use this?
```
    #: Only set this directly by changing the cls.escape
    _escape = None
```
However, property(escape) has no setter. Or am I missing something?


Also, could you please point me to the place where it is used? My question would be, does it apply NoEscape iteratively on all objects in ```self.data``` or only on string items of it?

Would appreciate if you helped me out here. Do not want to reinvent anything here.

**Comment by dgrigonis** [comment]
Ok, I looked at ```self._escape```. Correct me if I am wrong, but it is not used by any lower level latex generation as is only to be used directly from environments and commands!???

If this is the case, I think ```self._escape``` needs a setter of ```self.escape``` property.

**Comment by JelteF** [comment]
Hmm, that seems to be a bug indeed as well.

**Comment by JelteF** [comment]
Right now it is used here: https://github.com/JelteF/PyLaTeX/blob/master/pylatex/base_classes/containers.py#L68-L69

And _default_escape is set here: https://github.com/JelteF/PyLaTeX/blob/2c7c77b8ec13890ad5457714be417aad2ef8b299/pylatex/base_classes/command.py

**Comment by dgrigonis** [comment]
Ok, issued PR for setter:
[https://github.com/JelteF/PyLaTeX/pull/164](https://github.com/JelteF/PyLaTeX/pull/164)

Once it's merged, will make use of it in this PR.

**Comment by dgrigonis** [comment]
Ok, looked at it again and I am not sure I can use it in a way that its already been used.

Reason being that `escape` argument of `dumps_list` controls the whole list escaping.

In this case say you have:
```
Math(data=[Matrix(M), 'string \\'])
```
You do not want to affect Matrix behavior, only string as user/developer should make sure that matrix is correctly escaped before using it.

I have made use of `escape` property, but not in the way it was probably intended. But still a bit nicer than it was, maybe...

**Comment by dgrigonis** [comment]
Will edit changelog once we are happy with this PR.

**Comment by JelteF** [comment]
What's the advantage you gain from having the math environment without enclosing delimiters? I see you use it in the new Alignat environment. Am I missing something or do you not actually need the Math wrapper there? As far as I can tell you  could just append directly to `agn` in your example, instead of using the add_math method.

**Comment by dgrigonis** [comment]
Couple of things:

 1. `agn.append()` can only append one item at the time. Thus to add 3 matrices to 1 equation line, you need to append 3 times. Or of course you can do: `agn.data += [obj1, obj2], but this is user resorting to internals due to an absence of a good interface.
 2. The second reason is exactly why I wanted "no escape" option - that it is easy to write an equation in one line so that all strings are non-escaped without wrapping them in `NoEscape()`.

**Comment by dgrigonis** [comment]
Ok, I gave a thought one more time. Options are these:
1. Decouple `Alignat` from `Math`. Then there are several ways to achieve the functionality I have implemented.
2. Leave it as it is: the main advantage is that there will be no code duplication for keyword args in `Math` - whatever you implement in `Math` is readily available in `Alignat` via `add_math` method.

I prefer 2. Thus raw argument is useful - this will allow Math to be used by other math environments if implemented in the future via the same way. Furthermore, if user wants to define his/her own environment on the spot and use `Math` functionality, then all what needs to be done is `raw=True`. I think this allows healthy flexibility.

**Comment by JelteF** [comment]
1. You should be able to use `agn.extend()` as a nicer way.
2. I can understand your need for a container that doesn't do anything, except for making sure its components are not escaped. However, I think this might call for a seperate `NoEscapeContainer()` instead.

**Comment by JelteF** [comment]
> Ok, looked at it again and I am not sure I can use it in a way that its already been used.
> 
> Reason being that escape argument of dumps_list controls the whole list escaping.
> 

What is the problem with dumps_list handling the escaping?

**Comment by JelteF** [comment]
The NoEscapeContainer would look something like this:

```python
NoEscapeContainer(Container):
     _default_escape = False
```

**Comment by dgrigonis** [comment]
Ok, to answer your doubts:

 1. `dumps_list` handles escaping well. Apologies, my bad
 2. Used self.escape flag appropriately now
 3. `escape` and `raw` flags for `Math` container are useful even without `Alignat`
 4. Still think it is appropriate to use `Math` environment via add_math method

**Comment by JelteF** [comment]
I've thought about this again and I think the `add_math` is not needed if we design it a bit different. For inheriting of functionality subclass/mixin would be better, also right now `Math` doesn't have any special functionality. The add_math usage we have now in the example can be replaced with, `append` or `extend` list methods. The raw argument for math is then also not needed.

I can understand that the escape keyword for math is useful, but we cannot let it default to True. This would break backward compatibility. Could you make it default to None?  Also for Alignat for consistency.

File: pylatex/math.py [hunk 1]
```diff
@@ -38,10 +89,12 @@ def dumps(self):
...
         """
-
-        if self.inline:
+        if self.raw:
+            return self.dumps_content()
+        elif self.inline:
             return '$' + self.dumps_content() + '$'
-        return '\\[%\n' + self.dumps_content() + '%\n\\]'
+        else:
+            return '\\[%\n' + self.dumps_content() + '%\n\\]'
```
Comment (JelteF) [inline pylatex/math.py]: Can you remove this else, I like the previous style better.

**Comment by dgrigonis** [comment]
`What about escape=True` then? `None` means that it is undefined by the implementation.

**Comment by JelteF** [comment]
with escape `None` the value from `_default_escape` is used instead. The rest looks good now, so I'll merge it and change the defaults myself.

**Pull request merged**